analysis/R/assoc.R

```diff
@@ -0,0 +1,278 @@
+  params2 <- params
+  params2$k <- (params$k ** 2) * 4
+  
+  # Prune candidates
+  fit <- lapply(1:2, function(i)
+    Decode(counts[[i + 1]],
+           map[[i]]$rmap,
+           params, quick = FALSE)$fit)
+  
+  found_strings = list(fit[[1]][,"string"], fit[[2]][,"string"])
+  
+  if (length(found_strings[[1]]) == 0 || length(found_strings[[2]]) == 0) {
+    stop("No strings found in 1-way marginal.")
+  }
+  
+  # Combine maps to feed into Decode2Way
+  # Prune first to found_strings from Decode on 1-way counts
+  pruned <- lapply(1:2, function(i)
+    lapply(map[[i]]$map, function(z) z[,found_strings[[i]], drop = FALSE]))
+  crmap <- CombineMaps(pruned[[1]], pruned[[2]])$crmap
+  marginal <- Decode2Way(counts[[1]], crmap, params2, fit = fit)$fit
+  
+  # Reconstruct 2-way table from marginals
+  ed <- matrix(0, nrow = length(found_strings[[1]]), ncol = length(found_strings[[2]]))
+  colnames(ed) <- found_strings[[2]]
+  rownames(ed) <- found_strings[[1]]
+  for (cols in found_strings[[2]]) {
+    for (rows in found_strings[[1]]) {
+      ed[rows, cols] <- marginal[paste(rows, cols, sep = "x"), "Estimate"]
+    }
+  }
+  ed[is.na(ed)] <- 0
+  ed[ed<0] <- 0
+  
+  time_taken <- proc.time() - ptm
+  print("Two Way Algorithm Results")
+  print(ed[order(-rowSums(ed)), order(-colSums(ed))])
+  if(inp$time == TRUE)
+    print(time_taken)
+}
+
+EMAlg <- function(inp) {
+  ptm <- proc.time()
+  params <- ReadParameterFile(inp$params)
+  # Ensure sufficient maps as required by number of vars
+  stopifnot(inp$numvars == length(inp$maps))
+  # Correct map from ReadMapFile() for assoc analysis
+  map <- lapply(inp$maps, function(o)
+    CorrectMapForAssoc(LoadMapFile(o, params = params),
+                       params = params))
+  
+  # For BASIC only
+  m1 <- lapply(1:params$m, function(z) {
+    m <- sparseMatrix(c(1), c(2), dims = c(1, 2))
+    colnames(m) <- c("FALSE", "TRUE")
+    m
+  })
+  m2 <- sparseMatrix(1:params$m, rep(2, params$m))
+  colnames(m2) <- colnames(m1[[1]])
+  map[[2]]$map <- m1
+  map[[2]]$rmap <- m2
+  
+  # Reports must be of the format
+  #     client name, cohort no, rappor bitstring 1, rappor bitstring 2, ...
+  reportsObj <- read.csv(inp$reports,
+                         colClasses = c("character", "integer",
+                                        rep("character", inp$numvars)),
+                         header = TRUE)
+  # Ignore the first column
+  reportsObj <- reportsObj[,-1]
+  
+  params = list(params, params)
+  params[[2]]$k = 1
+  
+  # Parsing reportsObj
+  # ComputeDistributionEM allows for different sets of cohorts
+  # for each variable. Here, both sets of cohorts are identical
+  co <- as.list(reportsObj[1])[[1]]
+  co <- co + 1  # 1 indexing
+  cohorts <- rep(list(co), inp$numvars)
+  # Parse reports from reportObj cols 2, 3, ...
+  reports <- lapply(1:inp$numvars, function(x) as.list(reportsObj[x + 1]))
+  
+  # Split strings into bit arrays (as required by assoc analysis)
```

Comment:
FYI this is assuming that [[1]] is the string var, which will probably change.

Also I think it's more accurate to say "Split ASCII bit strings into a array of numerics?"  It might be more efficient in R does use as.integer instead of as.numeric?  If the only values are 0 or 1, then that seems to make more sense.